### PR TITLE
IOS-358: Port framework to OS X plarform

### PR DESCRIPTION
**Pull request by lukwol**
#### Description
- Moved iOS specific configurations to separate files.
- Added OS X specific configurations.
- Created OS X specific framework and test targets.
- Created `PNGImageConverter` to separate per platform image converting logic.
- Used `ImageType` typealias instead of platform specific `UIImage` and `NSImage`.
- Updated playground and tests to work with both platforms.
#### JIRA Link

https://netguru.atlassian.net/browse/IOS-358
#### Additional Notes
- I couldn't find better name for `ImageType` typealias.
  I was trying to use `Image` typealias instead of `ImageType`, but it won't work with `AnnotationRequest.Image`.
- There is no separate spec file for `PNGImageConverter`, because it's functionality is already tested in `Base64ImageEncoderSpec`.
#### Alternatives considered
- Creating `ImageType` protocol instead of typealias, and conform `UIImage` and `NSImage` to this protocol.
- Adding `PNGRepresentation` computed property in `ImageType` extension.
  I've decided not to make public function `PNGRepresentation` in extension and encapsulated converting to data logic in separate image converter, because I didn't want to add another method in image extension for the framework users.
  I choosed typealias instead of protocol, because `NSImage` and `UIImage` have similar interfaces, so I could call same methods and initializers without any additional work. The only problem was with converting to data, which was solved by separate struct.

### CODE REVIEW

File: Picguard.podspec [hunk 1]
```diff
@@ -10,15 +10,15 @@ Pod::Spec.new do |spec|
   # Description
 
   spec.name = 'Picguard'
-  spec.version = '0.2'
+  spec.version = '0.3'
```
Comment (lukwol) [inline Picguard.podspec]: I wonder if changes in podspec should be included in this PR. What do you think @akashivskyy ?

File: Configuration/Tests-Debug-OS X.xcconfig [hunk 2]
```diff
@@ -0,0 +1,9 @@
+//
+// Tests-Debug-OS X.xcconfig
+//
+// Copyright (c) 2016 Netguru Sp. z o.o. All rights reserved.
+// Licensed under the MIT License.
+//
+
+#include "Carthage/Checkouts/xcconfigs/Mac OS X/Mac-Application.xcconfig"
+#include "Tests-Debug.xcconfig"
```
Comment (akashivskyy) [inline Configuration/Tests-Debug-OS X.xcconfig]: Lots of missing newlines

File: Configuration/Framework-Debug-OS X.xcconfig [hunk 3]
```diff
@@ -0,0 +1,9 @@
+//
+// Framework-Debug-OS X.xcconfig
```
Comment (akashivskyy) [inline Configuration/Framework-Debug-OS X.xcconfig]: Let's be consistent with ResponseDetective and keep `[Subject]-[Platform]-[Configuration]` format, and no space between `OS` and `X`.

https://github.com/netguru/ResponseDetective/tree/develop/Configuration

File: Configuration/Framework-OSX-Debug.xcconfig [hunk 4]
```diff
@@ -0,0 +1,9 @@
+//
+// Framework-Debug-OS X.xcconfig
```
Comment (akashivskyy) [inline Configuration/Framework-OSX-Debug.xcconfig]: You forgot to change the header comments

File: Picguard.xcodeproj/project.pbxproj [hunk 5]
```diff
@@ -641,32 +902,83 @@
...
+			baseConfigurationReference = 5C1BD71D1CE6121700CB6364 /* Framework-OSX-Release.xcconfig */;
+			buildSettings = {
+			};
+			name = Release;
+		};
+		5C1BD73A1CE613AA00CB6364 /* Debug */ = {
+			isa = XCBuildConfiguration;
+			baseConfigurationReference = 5C1BD71E1CE6123600CB6364 /* Tests-OSX-Debug.xcconfig */;
+			buildSettings = {
+				CLANG_WARN_BOOL_CONVERSION = YES;
```
Comment (akashivskyy) [inline Picguard.xcodeproj/project.pbxproj]: Oh, hai build settings in pbxproj! 🙈

File: Sources/APIClient.swift [hunk 6]
```diff
@@ -49,11 +46,9 @@ public final class APIClient: APIClientType {
 	///       a session with default configuration.
 	public init(
 		APIKey: String,
-		encoder: ImageEncoding,
 		session: NSURLSession = NSURLSession(configuration: NSURLSessionConfiguration.defaultSessionConfiguration())
 	) {
 		self.APIKey = APIKey
-		self.encoder = encoder
```
Comment (akashivskyy) [inline Sources/APIClient.swift]: You didn't remove `encoder` from doc comment

File: Sources/APIClient.swift [hunk 7]
```diff
@@ -49,11 +46,9 @@ public final class APIClient: APIClientType {
 	///       a session with default configuration.
 	public init(
 		APIKey: String,
-		encoder: ImageEncoding,
```
Comment (akashivskyy) [inline Sources/APIClient.swift]: You didn't remove `encoder` from doc comment

File: Sources/AnnotationRequest.swift [hunk 8]
```diff
@@ -93,28 +91,28 @@ public struct AnnotationRequest {
...
 		case URL(String)
 
-		/// UIImage representation of image.
-		case Image(UIImage)
+		/// Platform specyfic ImageType representation of image.
+		case Image(Base64EncodableImage)
 
-		/// Data representation of image.
-		case Data(NSData)
+		/// Base 64 encoded string representation of image data.
```
Comment (akashivskyy) [inline Sources/AnnotationRequest.swift]: Base64 is a term, written without space

File: Sources/AnnotationRequest.swift [hunk 9]
```diff
@@ -93,28 +91,28 @@ public struct AnnotationRequest {
 		/// The Google Cloud Storage URI to the image.
 		case URL(String)
 
-		/// UIImage representation of image.
-		case Image(UIImage)
+		/// Platform specyfic ImageType representation of image.
```
Comment (akashivskyy) [inline Sources/AnnotationRequest.swift]: Typo: `specific`

File: Sources/AnnotationRequest.swift [hunk 10]
```diff
@@ -93,28 +91,28 @@ public struct AnnotationRequest {
...
+		case Base64String(String)
 
 		// MARK: JSON Representation
 
 		/// JSON dictionary representation of `Image`.
 		///
-		/// - Throws: Rethrows any errors thrown by `ImageEncoding` when encoder
-		///   fails to encode image to data.
+		/// - Throws: `Base64EncodableImageError.UnsupportedBitmapData` error,
+		///   when fails to get base64 encoded string representation of image data.
```
Comment (akashivskyy) [inline Sources/AnnotationRequest.swift]: when who fails

File: Sources/AnnotationRequest.swift [hunk 11]
```diff
@@ -213,9 +211,13 @@ public func == (lhs: AnnotationRequest.Image, rhs: AnnotationRequest.Image) -> B
...
 			return lhsURL == rhsURL
 		case let (.Image(lhsImage), .Image(rhsImage)):
-			return UIImagePNGRepresentation(lhsImage) == UIImagePNGRepresentation(rhsImage)
-		case let (.Data(lhsData), .Data(rhsData)):
-			return lhsData == rhsData
+			do {
+				return try lhsImage.base64EncodedStringRepresentation() == rhsImage.base64EncodedStringRepresentation()
+			} catch {
+				return false
+		}
```
Comment (akashivskyy) [inline Sources/AnnotationRequest.swift]: That `}` ran away

File: Sources/Base64EncodableImage.swift [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+
+
+/// Describes image with base64 encoded string representation
+/// to be used with Google Cloud Vision API.
+public protocol Base64EncodableImage {
+
+	/// Encodes the image into base64 string representation.
+	///
+	/// - Throws: Errors from `Base64EncodableImageError.UnsupportedBitmapData` error,
+	///   when fails to encode image.
```
Comment (akashivskyy) [inline Sources/Base64EncodableImage.swift]: Same as above.

File: Sources/NSImage.swift [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+//
+// Copyright (c) 2016 Netguru Sp. z o.o. All rights reserved.
+// Licensed under the MIT License.
+//
+
+import AppKit
+
+extension NSImage: Base64EncodableImage {
+
+	public func base64EncodedStringRepresentation() throws -> String {
```
Comment (akashivskyy) [inline Sources/NSImage.swift]:
```
- SeeAlso: Base64EncodableImage. base64EncodedStringRepresentation()
```

File: Sources/UIImage.swift [hunk 14]
```diff
@@ -0,0 +1,19 @@
...
+//
+// Copyright (c) 2016 Netguru Sp. z o.o. All rights reserved.
+// Licensed under the MIT License.
+//
+
+import UIKit
+
+extension UIImage: Base64EncodableImage {
+
+	public func base64EncodedStringRepresentation() throws -> String {
```
Comment (akashivskyy) [inline Sources/UIImage.swift]: Same as above (`SeeAlso`)

File: Tests/MockBase64EncodableImage.swift [hunk 15]
```diff
@@ -0,0 +1,15 @@
...
+//
+
+import Picguard
+
+final class MockBase64EncodableImage: Base64EncodableImage {
+
+	func base64EncodedStringRepresentation() throws -> String {
+		return "fixture encoded image"
+	}
+}
```
Comment (akashivskyy) [inline Tests/MockBase64EncodableImage.swift]: Newline missing.

**Comment by akashivskyy** [comment]
I'll fix the build settings and configuration, you fix doc comments. ;)

File: Configuration/Framework-OSX-Debug.xcconfig (hunk 4 above)
Comment (lukwol) [inline Configuration/Framework-OSX-Debug.xcconfig]: Oh crap...

File: Picguard.xcodeproj/project.pbxproj (hunk 5 above)
Comment (lukwol) [inline Picguard.xcodeproj/project.pbxproj]: I'm pretty sure I did cmd+a delete combo...

File: Sources/APIClient.swift (hunk 7 above)
Comment (lukwol) [inline Sources/APIClient.swift]: Thanks for notice :+1:

File: Sources/AnnotationRequest.swift (hunk 8 above)
Comment (lukwol) [inline Sources/AnnotationRequest.swift]: :+1:

File: Sources/AnnotationRequest.swift (hunk 10 above)
Comment (lukwol) [inline Sources/AnnotationRequest.swift]: :+1:

File: Sources/NSImage.swift (hunk 13 above)
Comment (lukwol) [inline Sources/NSImage.swift]: I left blank space here and forgot to add this note :)

**Comment by lukwol** [comment]
Ok

**Comment by lukwol** [comment]
Whoa missclick on phone, sorry :)

**Comment by akashivskyy** [comment]
Ok my part is done. ;)

**Comment by lukwol** [comment]
Mine too.

**Comment by akashivskyy** [comment]
Looks good!

**Pull request merged**